letsencrypt-auto-source/rebuild_dependencies.py

```diff
@@ -0,0 +1,221 @@
+#     without pinned dependencies, and in respect with input authoritative requirements
+#   - `certbot plugins` is called to check we have an healthy environment
+#   - finally current set of dependencies is extracted out of the docker using pip freeze
+SCRIPT = """\
+#!/bin/sh
+set -e
+
+cd /tmp/certbot
+letsencrypt-auto-source/letsencrypt-auto --install-only -n
+PYVER=`/opt/eff.org/certbot/venv/bin/python --version 2>&1 | cut -d" " -f 2 | cut -d. -f1,2 | sed 's/\.//'`
+
+/opt/eff.org/certbot/venv/bin/python letsencrypt-auto-source/pieces/create_venv.py /tmp/venv "$PYVER" "1"
+
+/tmp/venv/bin/python letsencrypt-auto-source/pieces/pipstrap.py
+/tmp/venv/bin/pip install -e acme -e . -e certbot-apache -e certbot-nginx -c /tmp/requirements.txt
+/tmp/venv/bin/certbot plugins
+/tmp/venv/bin/pip freeze >> /tmp/workspace/results
+"""
+
+
+def _process_one_distribution(distribution, verbose):
+    print('===> Gathering dependencies for {0}.'.format(distribution))
+    workspace = tempfile.mkdtemp()
+    script = os.path.join(workspace, 'script.sh')
+    authoritative_requirements = os.path.join(workspace, 'requirements.txt')
+    cid_file = os.path.join(workspace, 'cid')
+
+    try:
+        with open(script, 'w') as file_h:
+            file_h.write(SCRIPT)
+        os.chmod(script, 0o755)
+
+        with open(authoritative_requirements, 'w') as file_h:
+            file_h.write('\n'.join(['{0}=={1}'.format(package, version)
+                                    for package, version in AUTHORITATIVE_CONSTRAINTS.items()]))
+
+        command = ['docker', 'run', '--rm', '--cidfile', cid_file,
+                   '-v', '{0}:/tmp/certbot'.format(CERTBOT_REPO_PATH),
+                   '-v', '{0}:/tmp/workspace'.format(workspace),
+                   '-v', '{0}:/tmp/requirements.txt'.format(authoritative_requirements),
+                   distribution, '/tmp/workspace/script.sh']
+        sub_stdout = sys.stdout if verbose else subprocess.PIPE
+        sub_stderr = sys.stderr if verbose else subprocess.STDOUT
+        process = subprocess.Popen(command, stdout=sub_stdout, stderr=sub_stderr, universal_newlines=True)
+        stdoutdata, _ = process.communicate()
+
+        if process.returncode:
+            if stdoutdata:
+                sys.stderr.write('Output was:\n{0}'.format(stdoutdata))
+            raise RuntimeError('Error while gathering dependencies for {0}.'.format(distribution))
+
+        with open(os.path.join(workspace, 'results'), 'r') as file_h:
+            return file_h.read()
+    finally:
+        if os.path.isfile(cid_file):
+            with open(cid_file, 'r') as file_h:
+                cid = file_h.read()
+            subprocess.call(['docker', 'kill', cid], stdout=None, stderr=None)
+        shutil.rmtree(workspace)
+
+
+def _insert_results(dependencies_map, results, distribution):
+    refined_results = []
+    for result in results.split(os.linesep):
+        match = re.match(r'(.*)==(.*)', result)
```

Comment:
Maybe speed doesn't matter here, but this is going to do quite a bit of backtracking: the `.*` will read all the way to the end of the string, find it can't match the pattern, then try every substring starting from the end all the way until it finally finds the "==". Perhaps excluding "=" from consideration would work, like `([^=]+)==([^=]+)`. (I've also changed `*` to `+`, since we do expect something to be there.)

Reply:
It is a more focused regex. I test it.